File: engine.py
"""
Signsbeat Multi-Input Mitochondrial Intelligence Engine
Fixed reference implementation — all 8 layers corrected.

Fixes applied vs PDF spec:
  1. Layer numbering: second "Layer 4" renamed to Layer 5 (Redox Resilience).
  2. Layer 3 (Adaptation Score) — implemented (was specified but absent from code).
  3. Layer 5 (Redox Resilience) — implemented (was missing).
  4. Layer 6 (Metabolic Flexibility) — implemented (was missing).
  5. layer4_composite_reserve_capacity: division-by-zero guard on buffered_threshold.
  6. layer2_validate_composite_response: explicit no-intervention branch (was a silent fall-through).
  7. layer8_calculate_composite_health: multi-challenge bonus only awarded when recovery
     is NOT suppressed; double-penalty for suppression under multi-challenge reduced.
"""

import numpy as np
from typing import Dict, Any, List


class SignsbeatMultiInputMitochondrialEngine:

    def __init__(self):
        self.intervention_registry = {
            "mitochondrial_support": [
                "sleep", "coq10", "pqq", "creatine", "magnesium",
                "omega-3", "nac", "glycine", "protein_optimization",
            ],
            "mitochondrial_challenge": [
                "fasting", "ketogenic_diet", "exercise", "hiit",
                "sauna", "cold_plunge", "hbot", "red_light_therapy",
            ],
        }

    # ── Layer 1 ───────────────────────────────────────────────────────────────
    def layer1_process_daily_mix(self, interventions: List[Dict[str, Any]]) -> Dict[str, Any]:
        support_items, challenge_items = [], []
        cumulative_challenge_load = 0.0

        for item in interventions:
            name_lower = item["name"].lower().replace(" ", "_")
            load = item.get("load_hours", 1.0)
            if name_lower in self.intervention_registry["mitochondrial_support"]:
                support_items.append(name_lower)
            elif name_lower in self.intervention_registry["mitochondrial_challenge"]:
                challenge_items.append(name_lower)
                cumulative_challenge_load += load

        return {
            "support_count": len(support_items),
            "challenge_count": len(challenge_items),
            "cumulative_challenge_load": cumulative_challenge_load,
            "has_support": len(support_items) > 0,
            "has_challenge": len(challenge_items) > 0,
        }

    # ── Layer 2 ───────────────────────────────────────────────────────────────
    def layer2_validate_composite_response(
        self,
        mix_meta: Dict[str, Any],
        metrics_delta: Dict[str, float],
    ) -> Dict[str, Any]:
        pro_recovery_delta = metrics_delta.get("pro_recovery", 0.0)
        is_persistent_stress = metrics_delta.get("persistent_stress", False)

        # FIX: explicit no-intervention branch (original fell through silently)
        if not mix_meta["has_challenge"] and not mix_meta["has_support"]:
            return {"status": "Baseline — No Active Intervention", "is_positive": True}

        if mix_meta["has_challenge"] and mix_meta["has_support"]:
            if is_persistent_stress:
                return {
                    "status": "Critical Limitation: Persistent Stress Despite Support Buffering",
                    "is_positive": False,
                }
            if pro_recovery_delta > 1.5:
                return {"status": "Optimized Co-Intervention Rebound", "is_positive": True}
            return {"status": "Buffered Challenge — Monitor Trend", "is_positive": True}

        if mix_meta["has_challenge"] and not mix_meta["has_support"]:
            if not is_persistent_stress and pro_recovery_delta > 0:
                return {"status": "Positive Adaptation Rebound", "is_positive": True}
            if is_persistent_stress:
                return {"status": "Potential Mitochondrial Limitation", "is_positive": False}
            return {"status": "Challenge Tolerated — No Rebound Yet", "is_positive": False}

        # Support only, no challenge
        if pro_recovery_delta > 0 and metrics_delta.get("pro_stress", 0.0) <= 0:
            return {"status": "Positive Pure Support Response", "is_positive": True}
        return {"status": "Support Active — Insufficient Response Signal", "is_positive": False}

    # ── Layer 3 (FIX: was specified but not implemented) ──────────────────────
    def layer3_adaptation_score(
        self,
        previous_pro_recovery: float,
        current_pro_recovery: float,
        stress_generated: float,
    ) -> float:
        """Adaptation Score = Recovery Gain / Stress Generated (centred to 0–100)."""
        recovery_gain = current_pro_recovery - previous_pro_recovery
        if stress_generated <= 0:
            return 100.0 if recovery_gain > 0 else 50.0
        raw_ratio = recovery_gain / stress_generated
        return float(np.clip(raw_ratio * 50 + 50, 0.0, 100.0))

    # ── Layer 4: Reserve Capacity ─────────────────────────────────────────────
    def layer4_composite_reserve_capacity(
        self,
        cumulative_load: float,
        base_threshold: float,
        support_count: int,
    ) -> float:
        # FIX: guard against zero base_threshold
        effective_base = base_threshold if base_threshold > 0 else 16.0
        buffered_threshold = effective_base * (1.0 + support_count * 0.10)
        if cumulative_load > buffered_threshold:
            decrement = min(1.0, (cumulative_load - buffered_threshold) / buffered_threshold)
            return float(max(0.0, 1.0 - decrement))
        return 1.0

    # ── Layer 5: Redox Resilience (FIX: was mislabelled Layer 4 and not coded) ──
    def layer5_redox_resilience(
        self,
        hrv_trend: float,
        persistent_stress: bool,
        pro_recovery_delta: float,
    ) -> Dict[str, Any]:
        score = 50.0
        if hrv_trend > 0:         score += 20.0
        if not persistent_stress: score += 15.0
        if pro_recovery_delta > 0: score += 15.0
        if persistent_stress:      score -= 30.0
        if hrv_trend < -1:         score -= 20.0
        clamped = float(np.clip(score, 0.0, 100.0))
        if clamped >= 70:
            phenotype = "Healthy — Rapid Over-Compensatory Rebound"
        elif clamped >= 40:
            phenotype = "Transitional — Partial ROS Clearance"
        else:
            phenotype = "Dysfunctional — Sustained Unmitigated Stress"
        return {"score": clamped, "phenotype": phenotype}

    # ── Layer 6: Metabolic Flexibility (FIX: was specified but not implemented) ──
    def layer6_metabolic_flexibility(
        self,
        substrate_shifts: int,
        glucose_volatility: float,
        hrv_trend: float,
    ) -> Dict[str, Any]:
        score = 50.0
        score += min(substrate_shifts * 10, 30)
        score -= min(glucose_volatility * 5, 30)
        if hrv_trend > 0: score += 10.0
        clamped = float(np.clip(score, 0.0, 100.0))
        flexibility_class = (
            "Flexible" if clamped >= 70
            else "Moderately Flexible" if clamped >= 40
            else "Inflexible"
        )
        return {"score": clamped, "flexibility_class": flexibility_class}

    # ── Layer 7 ───────────────────────────────────────────────────────────────
    def layer7_calculate_composite_dysfunction(
        self,
        risk_factors: Dict[str, Any],
        mix_meta: Dict[str, Any],
        validation_status: str,
    ) -> float:
        score = 0.0
        if risk_factors.get("persistent_pro_stress", False): score += 30.0
        if risk_factors.get("declining_hrv", False):         score += 20.0
        if risk_factors.get("poor_sleep", False):            score += 20.0
        if mix_meta["has_support"] and "Critical Limitation" in validation_status:
            score += 30.0
        return float(np.clip(score, 0.0, 100.0))

    # ── Layer 8 ───────────────────────────────────────────────────────────────
    def layer8_calculate_composite_health(
        self,
        metrics: Dict[str, Any],
        mix_meta: Dict[str, Any],
    ) -> float:
        pos_score = 0.0
        if metrics.get("rising_pro_recovery", False): pos_score += 25.0
        if metrics.get("improved_hrv", False):        pos_score += 20.0
        if metrics.get("improved_sleep", False):      pos_score += 15.0
        if metrics.get("faster_recovery", False):     pos_score += 20.0

        # FIX: bonus only when recovery is NOT suppressed
        if (
            mix_meta["challenge_count"] >= 2
            and metrics.get("faster_recovery", False)
            and not metrics.get("recovery_suppression", False)
        ):
            pos_score += 20.0

        neg_penalty = 0.0
        if metrics.get("persistent_pro_stress", False): neg_penalty += 35.0
        if metrics.get("recovery_suppression", False):   neg_penalty += 35.0

        # FIX: reduced from 20 to 10 to avoid excessive stacking
        if (
            mix_meta["challenge_count"] >= 2
            and metrics.get("recovery_suppression", False)
            and not metrics.get("persistent_pro_stress", False)
        ):
            neg_penalty += 10.0

        return float(np.clip(pos_score - neg_penalty, 0.0, 100.0))

    # ── Orchestrator ──────────────────────────────────────────────────────────
    def evaluate_daily_mitochondrial_health(self, daily_profile: Dict[str, Any]) -> Dict[str, Any]:
        interventions = daily_profile.get("interventions", [])
        metrics_delta = daily_profile.get("metrics_delta", {})

        mix_meta = self.layer1_process_daily_mix(interventions)
        validation = self.layer2_validate_composite_response(mix_meta, metrics_delta)

        prev_rec = daily_profile.get("previous_pro_recovery", 0.0)
        curr_rec = daily_profile.get("current_pro_recovery", prev_rec + metrics_delta.get("pro_recovery", 0.0))
        adaptation = self.layer3_adaptation_score(prev_rec, curr_rec, abs(metrics_delta.get("pro_stress", 0.0)))

        reserve_cap = self.layer4_composite_reserve_capacity(
            cumulative_load=mix_meta["cumulative_challenge_load"],
            base_threshold=daily_profile.get("base_threshold_hours", 16.0),
            support_count=mix_meta["support_count"],
        )

        hrv = daily_profile.get("hrv_trend", 0.0)
        sleep = daily_profile.get("sleep_trend", 0.0)
        redox = self.layer5_redox_resilience(hrv, metrics_delta.get("persistent_stress", False), metrics_delta.get("pro_recovery", 0.0))
        metabolic = self.layer6_metabolic_flexibility(daily_profile.get("substrate_shifts", 1), daily_profile.get("glucose_volatility", 0.0), hrv)

        risk_factors = {
            "persistent_pro_stress": metrics_delta.get("persistent_stress", False),
            "declining_hrv": hrv < 0,
            "poor_sleep": sleep < 0,
        }
        dysfunction_prob = self.layer7_calculate_composite_dysfunction(risk_factors, mix_meta, validation["status"])

        health_inputs = {
            "rising_pro_recovery": metrics_delta.get("pro_recovery", 0.0) > 0,
            "improved_hrv": hrv > 0,
            "improved_sleep": sleep > 0,
            "faster_recovery": validation["is_positive"],
            "recovery_suppression": metrics_delta.get("pro_recovery", 0.0) < -0.5,
            "persistent_pro_stress": metrics_delta.get("persistent_stress", False),
        }
        final_health = self.layer8_calculate_composite_health(health_inputs, mix_meta)

        return {
            "Profile": f"Challenges: {mix_meta['challenge_count']} | Supports: {mix_meta['support_count']}",
            "Composite Load": mix_meta["cumulative_challenge_load"],
            "Response Status": validation["status"],
            "Adaptation Score (L3)": f"{adaptation:.1f}%",
            "Reserve Capacity Score (L4)": f"{round(reserve_cap * 100, 1)}%",
            "Redox Resilience Score (L5)": f"{redox['score']:.1f}% — {redox['phenotype']}",
            "Metabolic Flexibility Score (L6)": f"{metabolic['score']:.1f}% — {metabolic['flexibility_class']}",
            "Dysfunction Probability (L7)": f"{dysfunction_prob:.1f}%",
            "Mitochondrial Health Score (L8)": f"{final_health:.1f}%",
        }


# ── Quick smoke test ──────────────────────────────────────────────────────────
if __name__ == "__main__":
    engine = SignsbeatMultiInputMitochondrialEngine()
    profile = {
        "interventions": [
            {"name": "exercise", "load_hours": 1.5},
            {"name": "hiit", "load_hours": 0.5},
            {"name": "sleep", "load_hours": 8},
            {"name": "magnesium", "load_hours": 1},
            {"name": "omega-3", "load_hours": 1},
        ],
        "metrics_delta": {"pro_recovery": 2.1, "pro_stress": 0.3, "persistent_stress": False},
        "hrv_trend": 3.5,
        "sleep_trend": 0.8,
        "base_threshold_hours": 16,
        "previous_pro_recovery": 5.0,
        "current_pro_recovery": 7.1,
        "substrate_shifts": 2,
        "glucose_volatility": 0.5,
    }
    result = engine.evaluate_daily_mitochondrial_health(profile)
    for k, v in result.items():
        print(f"{k:40s} {v}")

File: lib/engine.ts
// Signsbeat Multi-Input Mitochondrial Intelligence Engine
// TypeScript — exact port of the reference Python spec
// Terminology: "intervention" replaced with "ActionName" throughout

export interface ActionName {
  name: string;
  load_hours?: number;
}

export interface MetricsDelta {
  pro_recovery?: number;      // delta vs previous day (positive = improving)
  pro_mild_stress?: number;   // delta vs previous day for Pro_MildStress
  pro_stress?: number;        // current pro_stress value or delta
  persistent_stress?: boolean;
}

export interface DailyProfile {
  action_names: ActionName[];          // was: interventions
  metrics_delta: MetricsDelta;
  base_threshold_hours?: number;       // default 16.0
  hrv_trend?: number;
  sleep_trend?: number;
  // Raw Pro_State values for display
  pro_positive?: number;
  pro_recovery?: number;
  pro_mild_stress?: number;
  pro_stress?: number;
}

export interface EngineResult {
  // Layer outputs (matches Python return dict, with ActionName terminology)
  action_name_profile: string;            // was: "Intervention Profile"
  cumulative_challenge_load: string;
  dynamic_response_status: string;
  reserve_capacity_score: string;
  dysfunction_probability: string;
  final_mito_health_score: string;

  // Extended fields for UI
  is_positive: boolean;
  adaptation_score: number;
  support_action_names: string[];
  challenge_action_names: string[];
  unknown_action_names: string[];
  // ActionNames not in registry but reclassified via observed Pro_State response
  inferred_challenge_names: string[];   // promoted Pro_Recovery → inferred Challenge
  inferred_support_names: string[];     // promoted Pro_MildStress → inferred Support
  pro_state_values: {
    pro_positive: number;
    pro_recovery: number;
    pro_mild_stress: number;
    pro_stress: number;
  };
  layer_breakdown: {
    support_count: number;
    challenge_count: number;
    cumulative_load: number;
    buffered_threshold: number;
    reserve_capacity: number;
    dysfunction_score: number;
    health_score: number;
  };
}

// ── ActionName Registry (Layer 1) ─────────────────────────────────────────────

const ACTION_NAME_REGISTRY = {
  mitochondrial_support: new Set([
    "sleep", "coq10", "pqq", "creatine", "magnesium",
    "omega-3", "omega3", "nac", "glycine", "protein_optimization",
    "protein", "zinc", "vitamin_d", "vitamin_c", "ashwagandha",
    "l-theanine", "ltheanine", "melatonin", "berberine", "nmn",
    "nad", "probiotics", "fish_oil", "multivitamin",
    "rest", "recovery_session", "meditation", "breathwork", "massage",
  ]),
  mitochondrial_challenge: new Set([
    "fasting", "intermittent_fasting", "ketogenic_diet", "ketogenic", "keto",
    "exercise", "hiit", "resistance", "strength", "weight_training",
    "cardio", "aerobic", "running", "cycling", "swimming",
    "zone_2", "sprints", "crossfit", "vo2max", "tempo",
    "sauna", "infrared_sauna", "cold_plunge", "cold_water", "cwi",
    "hbot", "hyperbaric", "red_light_therapy", "red_light",
  ]),
};

export function classifyActionName(name: string): "support" | "challenge" | "unknown" {
  const key = name.toLowerCase().trim().replace(/[\s\-\/]+/g, "_");

  if (ACTION_NAME_REGISTRY.mitochondrial_support.has(key)) return "support";
  if (ACTION_NAME_REGISTRY.mitochondrial_challenge.has(key)) return "challenge";

  // Partial / substring match
  const supportArr   = Array.from(ACTION_NAME_REGISTRY.mitochondrial_support);
  const challengeArr = Array.from(ACTION_NAME_REGISTRY.mitochondrial_challenge);
  for (const s of supportArr)   { if (key.includes(s) || s.includes(key)) return "support"; }
  for (const c of challengeArr) { if (key.includes(c) || c.includes(key)) return "challenge"; }

  return "unknown";
}

// ── Layer 1: ActionName Classification & Multi-Input Mixer ───────────────────
// Unknown ActionNames are reclassified via observed Pro_State promotion:
//   • Pro_Recovery delta > 0  → inferred Challenge (hormetic adaptation signal)
//   • Pro_MildStress delta > 0 → inferred Support   (mild modulation signal)
//   Exception: items already in the hormetic registry stay as Challenge regardless.

function layer1_process_daily_mix(
  action_names: ActionName[],
  pro_recovery_delta: number = 0,
  pro_mild_stress_delta: number = 0,
): {
  support_count: number;
  challenge_count: number;
  cumulative_challenge_load: number;
  has_support: boolean;
  has_challenge: boolean;
  support_names: string[];
  challenge_names: string[];
  unknown_names: string[];
  inferred_challenge_names: string[];
  inferred_support_names: string[];
} {
  const support_names:          string[] = [];
  const challenge_names:        string[] = [];
  const unknown_names:          string[] = [];
  const inferred_challenge_names: string[] = [];
  const inferred_support_names:   string[] = [];
  let cumulative_challenge_load = 0.0;

  for (const item of action_names) {
    const load = item.load_hours ?? 1.0;
    const cls  = classifyActionName(item.name);

    if (cls === "support") {
      support_names.push(item.name);
    } else if (cls === "challenge") {
      challenge_names.push(item.name);
      cumulative_challenge_load += load;
    } else {
      // Unknown — reclassify from observed Pro_State promotion
      if (pro_recovery_delta > 0) {
        // Pro_Recovery promoted → hormetic/challenge response inferred
        inferred_challenge_names.push(item.name);
        challenge_names.push(item.name);          // counted as challenge
        cumulative_challenge_load += load;
      } else if (pro_mild_stress_delta > 0) {
        // Pro_MildStress promoted → supportive/modulating response inferred
        inferred_support_names.push(item.name);
        support_names.push(item.name);            // counted as support
      } else {
        // No observable Pro_State signal — stays truly unknown
        unknown_names.push(item.name);
      }
    }
  }

  return {
    support_count: support_names.length,
    challenge_count: challenge_names.length,
    cumulative_challenge_load,
    has_support:   support_names.length > 0,
    has_challenge: challenge_names.length > 0,
    support_names,
    challenge_names,
    unknown_names,
    inferred_challenge_names,
    inferred_support_names,
  };
}

// ── Layer 2: Composite Response Validation ───────────────────────────────────

function layer2_validate_composite_response(
  mix_meta: ReturnType<typeof layer1_process_daily_mix>,
  metrics_delta: MetricsDelta,
): { status: string; is_positive: boolean } {
  const pro_recovery_delta  = metrics_delta.pro_recovery  ?? 0.0;
  const pro_stress_delta    = metrics_delta.pro_stress    ?? 0.0;
  const is_persistent_stress = metrics_delta.persistent_stress ?? false;

  // Challenge + Support deployed together
  if (mix_meta.has_challenge && mix_meta.has_support) {
    if (is_persistent_stress) {
      return { status: "Critical Limitation: Persistent Stress despite Support Buffering", is_positive: false };
    }
    if (pro_recovery_delta > 1.5) {
      return { status: "Optimized Co-ActionName Rebound", is_positive: true };
    }
    // Has both but response is moderate — still positive direction
    if (pro_recovery_delta > 0) {
      return { status: "Buffered Challenge — Positive Trend", is_positive: true };
    }
  }

  // Challenge only, no support
  if (mix_meta.has_challenge && !mix_meta.has_support) {
    if (!is_persistent_stress && pro_recovery_delta > 0) {
      return { status: "Positive Adaptation Rebound", is_positive: true };
    }
    if (is_persistent_stress) {
      return { status: "Potential Mitochondrial Limitation", is_positive: false };
    }
  }

  // Support only, no challenge
  if (mix_meta.has_support && !mix_meta.has_challenge) {
    if (pro_recovery_delta > 0 && pro_stress_delta <= 0) {
      return { status: "Positive Pure Support Response", is_positive: true };
    }
    return { status: "Support Active — Insufficient Response Signal", is_positive: false };
  }

  // No ActionNames at all
  if (!mix_meta.has_challenge && !mix_meta.has_support) {
    return { status: "No ActionName Data — Score is 0", is_positive: false };
  }

  return { status: "Investigate Dynamic Mismatch", is_positive: false };
}

// ── Layer 4: Composite Reserve Capacity ──────────────────────────────────────
// Every support ActionName expands total daily reserve capacity threshold by 10%

function layer4_composite_reserve_capacity(
  cumulative_load: number,
  base_threshold: number,
  support_count: number,
): { buffered_threshold: number; reserve_capacity: number } {
  const effective_base   = base_threshold > 0 ? base_threshold : 16.0;
  const buffered_threshold = effective_base * (1.0 + support_count * 0.10);

  if (cumulative_load > buffered_threshold) {
    const decrement = Math.min(1.0, (cumulative_load - buffered_threshold) / buffered_threshold);
    return { buffered_threshold, reserve_capacity: Math.max(0.0, 1.0 - decrement) };
  }
  return { buffered_threshold, reserve_capacity: 1.0 };
}

// ── Layer 7: Composite Dysfunction Probability ───────────────────────────────
// Elevates dysfunction probability if system fails to adapt despite heavy support stacks

function layer7_calculate_composite_dysfunction(
  risk_factors: { persistent_pro_stress: boolean; declining_hrv: boolean; poor_sleep: boolean },
  mix_meta: ReturnType<typeof layer1_process_daily_mix>,
  validation_status: string,
): number {
  let score = 0.0;

  if (risk_factors.persistent_pro_stress) score += 30.0;
  if (risk_factors.declining_hrv)         score += 20.0;
  if (risk_factors.poor_sleep)            score += 20.0;

  // High penalty: failing to recover even with support optimisation
  if (mix_meta.has_support && validation_status.includes("Critical Limitation")) {
    score += 30.0;
  }

  return Math.min(100.0, Math.max(0.0, score));
}

// ── Layer 8: Composite Health Score ──────────────────────────────────────────
// Aggregates multi-ActionName telemetry vectors into a final scaled probability

function layer8_calculate_composite_health(
  metrics: {
    rising_pro_recovery: boolean;
    improved_hrv: boolean;
    improved_sleep: boolean;
    faster_recovery: boolean;
    recovery_suppression: boolean;
    persistent_pro_stress: boolean;
  },
  mix_meta: ReturnType<typeof layer1_process_daily_mix>,
): number {
  let pos_score = 0.0;

  if (metrics.rising_pro_recovery) pos_score += 25.0;
  if (metrics.improved_hrv)        pos_score += 20.0;
  if (metrics.improved_sleep)      pos_score += 15.0;
  if (metrics.faster_recovery)     pos_score += 20.0;

  // Synergistic bonus for stacking multiple successfully tolerated ActionNames
  if (mix_meta.challenge_count >= 2 && metrics.faster_recovery) {
    pos_score += 20.0;
  }

  let neg_penalty = 0.0;

  if (metrics.persistent_pro_stress)  neg_penalty += 35.0;
  if (metrics.recovery_suppression)   neg_penalty += 35.0;

  // Extra penalty if multi-challenge stacking causes unmitigated collapse
  if (mix_meta.challenge_count >= 2 && metrics.recovery_suppression) {
    neg_penalty += 20.0;
  }

  return Math.min(100.0, Math.max(0.0, pos_score - neg_penalty));
}

// ── Main Orchestrator ─────────────────────────────────────────────────────────

export function evaluateDailyMitochondrialHealth(profile: DailyProfile): EngineResult {
  const action_names  = profile.action_names ?? [];
  const metrics_delta = profile.metrics_delta ?? {};

  // Layer 1 — classify all ActionNames
  // Pass Pro_State deltas so unknown ActionNames can be reclassified by observed response
  const pro_recovery_delta    = metrics_delta.pro_recovery    ?? 0;
  const pro_mild_stress_delta = metrics_delta.pro_mild_stress ?? 0;
  const mix_meta = layer1_process_daily_mix(action_names, pro_recovery_delta, pro_mild_stress_delta);

  // Layer 2 — validate composite physiological signatures
  const validation = layer2_validate_composite_response(mix_meta, metrics_delta);

  // Adaptation score: Recovery Gain / Stress Generated
  const adaptation_score =
    (metrics_delta.pro_recovery ?? 0) /
    Math.max(0.1, metrics_delta.pro_stress ?? 0.1);

  // Layer 4 — adaptive buffering equation
  const { buffered_threshold, reserve_capacity } = layer4_composite_reserve_capacity(
    mix_meta.cumulative_challenge_load,
    profile.base_threshold_hours ?? 16.0,
    mix_meta.support_count,
  );

  // Layer 7 — dysfunction risk
  const risk_factors = {
    persistent_pro_stress: metrics_delta.persistent_stress ?? false,
    declining_hrv:         (profile.hrv_trend ?? 0) < 0,
    poor_sleep:            (profile.sleep_trend ?? 0) < 0,
  };
  const dysfunction_score = layer7_calculate_composite_dysfunction(risk_factors, mix_meta, validation.status);

  // Layer 8 — final health probability
  const health_inputs = {
    rising_pro_recovery:  (metrics_delta.pro_recovery ?? 0) > 0,
    improved_hrv:         (profile.hrv_trend ?? 0) > 0,
    improved_sleep:       (profile.sleep_trend ?? 0) > 0,
    faster_recovery:      validation.is_positive,
    recovery_suppression: (metrics_delta.pro_recovery ?? 0) < -0.5,
    persistent_pro_stress: metrics_delta.persistent_stress ?? false,
  };
  const health_score = layer8_calculate_composite_health(health_inputs, mix_meta);

  return {
    // Python-equivalent output keys (ActionName terminology)
    action_name_profile:      `Challenges: ${mix_meta.challenge_count} | Supports: ${mix_meta.support_count}`,
    cumulative_challenge_load: `${mix_meta.cumulative_challenge_load.toFixed(1)} Factor-Hours`,
    dynamic_response_status:  validation.status,
    reserve_capacity_score:   `${(reserve_capacity * 100).toFixed(1)}%`,
    dysfunction_probability:  `${dysfunction_score.toFixed(1)}%`,
    final_mito_health_score:  `${health_score.toFixed(1)}%`,

    // UI extensions
    is_positive:              validation.is_positive,
    adaptation_score:         parseFloat(adaptation_score.toFixed(2)),
    support_action_names:     mix_meta.support_names.filter(n => !mix_meta.inferred_support_names.includes(n)),
    challenge_action_names:   mix_meta.challenge_names.filter(n => !mix_meta.inferred_challenge_names.includes(n)),
    unknown_action_names:     mix_meta.unknown_names,
    inferred_challenge_names: mix_meta.inferred_challenge_names,
    inferred_support_names:   mix_meta.inferred_support_names,
    pro_state_values: {
      pro_positive:    Math.round((profile.pro_positive    ?? 0) * 100),
      pro_recovery:    Math.round((profile.pro_recovery    ?? 0) * 100),
      pro_mild_stress: Math.round((profile.pro_mild_stress ?? 0) * 100),
      pro_stress:      Math.round((profile.pro_stress      ?? 0) * 100),
    },
    layer_breakdown: {
      support_count:     mix_meta.support_count,
      challenge_count:   mix_meta.challenge_count,
      cumulative_load:   mix_meta.cumulative_challenge_load,
      buffered_threshold,
      reserve_capacity,
      dysfunction_score,
      health_score,
    },
  };
}
